### PR TITLE
fix the inconsistent description of TopologyKey in PodAffinityTerm

### LINKED ISSUE #54276: podAffinityTerm.topologyKey is required
The field is marked as optional:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L2414

The comment also describes a use case for empty topologykey:
https://github.com/kubernetes/kubernetes/blame/master/staging/src/k8s.io/api/core/v1/types.go#L2410

But the validation rule says it's required: 
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/validation/validation.go#L2662

### PR DESCRIPTION
**Pull request by guangxuli**
**What this PR does / why we need it**:
Clarify the confusing of inconsistent description.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Just fix #54276
**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by guangxuli** [comment]
all tests are green.
@aveshagarwal @caesarxuchao @saad-ali PTAL.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 1]
```diff
@@ -2486,9 +2486,7 @@ type PodAffinityTerm struct {
 	// the labelSelector in the specified namespaces, where co-located is defined as running on a node
 	// whose value of the label with key topologyKey matches that of any node on which any of the
 	// selected pods is running.
-	// For PreferredDuringScheduling pod anti-affinity, empty topologyKey is interpreted as "all topologies"
-	// ("all topologies" here means all the topologyKeys indicated by scheduler command-line argument --failure-domains);
-	// for affinity and for RequiredDuringScheduling pod anti-affinity, empty topologyKey is not allowed.
+	// Empty topologyKey is not allowed.
 	// +optional
 	TopologyKey string `json:"topologyKey,omitempty" protobuf:"bytes,3,opt,name=topologyKey"`
```
Comment (aveshagarwal) [inline staging/src/k8s.io/api/core/v1/types.go]: i think you need to remove omitempty too unless there is a default value assigned. could you check it?

Comment (guangxuli) [reply staging/src/k8s.io/api/core/v1/types.go]: Agree, i didn't recognize that point,thanks.

**Comment by guangxuli** [comment]
@aveshagarwal ptal.

**Comment by guangxuli** [comment]
kindly ping @caesarxuchao @aveshagarwal

**Comment by aveshagarwal** [comment]
lgtm.

**Comment by guangxuli** [comment]
@smarterclayton or @liggit need your approve,thanks.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 2]
```diff
@@ -2486,11 +2486,9 @@ type PodAffinityTerm struct {
 	// the labelSelector in the specified namespaces, where co-located is defined as running on a node
 	// whose value of the label with key topologyKey matches that of any node on which any of the
 	// selected pods is running.
-	// For PreferredDuringScheduling pod anti-affinity, empty topologyKey is interpreted as "all topologies"
-	// ("all topologies" here means all the topologyKeys indicated by scheduler command-line argument --failure-domains);
-	// for affinity and for RequiredDuringScheduling pod anti-affinity, empty topologyKey is not allowed.
+	// Empty topologyKey is not allowed.
 	// +optional
```
Comment (caesarxuchao) [inline staging/src/k8s.io/api/core/v1/types.go]: Remove the `+optional`

Comment (caesarxuchao) [reply staging/src/k8s.io/api/core/v1/types.go]: Otherwise lgtm.

**Comment by guangxuli** [comment]
hi @caesarxuchao  do you fully sure this field is Mandatory?  A little confusing that in our internal api type definition, [topologyKey](https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L2240) is still optional.

**Comment by liggitt** [comment]
In all validation cases I could find, an API object that leaves it empty fails validation. I'm tentatively ok with this, assuming due diligence to sweep all places this field is used/validated to ensure this isn't tightening validation on any persisted objects

**Comment by guangxuli** [comment]
thanks @liggitt hmm, do you mean in this PR we could remove the `optional`? I have no plan to sweep all the places, just only confirm the api definition rules, i am not fully familiar with this area. :)

**Comment by guangxuli** [comment]
@caesarxuchao @aveshagarwal @liggitt PTAL. thanks.

**Comment by caesarxuchao** [comment]
@liggitt the tightening of the validation was done in https://github.com/kubernetes/kubernetes/pull/49976, this one is just updating the api description to reflect the truth. Do you have other concerns?

**Comment by liggitt** [comment]
Sgtm

**Comment by guangxuli** [comment]
Any comments for this PR?

**Pull request merged**